typescript/src/redemption.ts

```diff
@@ -0,0 +1,291 @@
+    mainUtxo.transactionHash
+  )
+
+  const mainUtxoWithRaw: UnspentTransactionOutput & RawTransaction = {
+    ...mainUtxo,
+    transactionHex: rawTransaction.transactionHex,
+  }
+
+  const redemptionRequests = await prepareRedemptionRequests(
+    bridge,
+    walletPrivateKey,
+    redeemerAddresses
+  )
+
+  const transaction = await createRedemptionTransaction(
+    walletPrivateKey,
+    mainUtxoWithRaw,
+    redemptionRequests,
+    witness
+  )
+
+  // Note that `broadcast` may fail silently (i.e. no error will be returned,
+  // even if the transaction is rejected by other nodes and does not enter the
+  // mempool, for example due to an UTXO being already spent).
+  await bitcoinClient.broadcast(transaction)
+}
+
+/**
+ * Prepares a list of redemption requests based on the provided redeemer
+ * addresses and the wallet key. Details of the redemption requests are fetched
+ * from the provided Bridge on-chain contract handle.
+ * @dev It is up to the caller of this function to ensure that each of the
+ *      addresses represent a valid pending redemption request in the Bridge
+ *      on-chain contract. An exception will be thrown if any of the addresses
+ *      (along with the wallet public key corresponding to the provided private
+ *      key) does not represent a valid pending redemption.
+ * @param bridge The handle to the Bridge on-chain contract
+ * @param walletPrivateKey The private key of the wallet in the WIF format
+ * @param redeemerAddresses The addresses that will be the recipients of the
+ *        redeemed Bitcoins
+ * @returns The list of redemption requests.
+ */
+async function prepareRedemptionRequests(
+  bridge: Bridge,
+  walletPrivateKey: string,
+  redeemerAddresses: string[]
+): Promise<RedemptionRequest[]> {
+  const walletKeyRing = createKeyRing(walletPrivateKey)
+  const walletPublicKey = walletKeyRing.getPublicKey().toString("hex")
+
+  const walletPubKeyHash = `0x${hash160
+    .digest(Buffer.from(walletPublicKey, "hex"))
+    .toString("hex")}`
+
+  const redemptionRequests: RedemptionRequest[] = []
+
+  for (const redeemerAddress of redeemerAddresses) {
+    const redeemerOutputScript = deriveOutputScript(redeemerAddress)
+
+    const pendingRedemption = await bridge.pendingRedemptions(
+      walletPubKeyHash,
+      redeemerOutputScript
+    )
+
+    if (pendingRedemption.requestedAt == 0) {
+      // The requested redemption does not exist among `pendingRedemptions`
+      // in the Bridge.
+      throw new Error(
+        "Provided redeemer address and wallet public key do not identify a pending redemption"
+      )
+    }
+
+    // TODO: Use the value of fee that was set in the Bridge (max fee) as the
+    // `txFee` for now.
+    // In the future allow the caller to propose the value of transaction fee.
+    // If the proposed transaction fee is smaller than the sum of fee shares from
+    // all the outputs then use the proposed fee and add the difference to outputs
+    // proportionally.
```

Comment:
Step 5: I agree with this TODO but I think that the `createRedemptionTransaction` function should be responsible for setting the fee. Let's move that TODO there.